pkg/session/syssession/session.go

```diff
@@ -0,0 +1,477 @@
+// it will return the internal SessionContext for further usage and a function to exit the operation.
+func (s *session) EnterOperation(caller sessionOwner) (SessionContext, func(), error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	seqStart := s.seqInc()
+	if err := s.checkOwner(caller, "EnterOperation error: "); err != nil {
+		s.reportError(
+			err.Error(),
+			zap.Uint64("seqStart", seqStart),
+			zap.String("caller", objectStr(caller)),
+		)
+		return nil, nil, err
+	}
+
+	s.inuse++
+	exit := false
+	return s.sctx, func() {
+		r := recover()
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		seqEnd := s.seqInc()
+		if exit {
+			s.reportError(
+				"EnterOperation error: multiple exits for the same operation",
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("caller", objectStr(caller)),
+			)
+			return
+		}
+
+		exit = true
+		s.inuse--
+		if s.owner != caller {
+			s.reportError(
+				"ExitOperation error: session owner transferred when executing operation",
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("caller", objectStr(caller)),
+			)
+		}
+		if r != nil {
+			logutil.BgLogger().Error(
+				"EnterOperation error: panic occurs, avoid to use session again",
+				zap.Any("panic", r),
+				zap.Uint64("seqStart", seqStart),
+				zap.Uint64("seqEnd", seqEnd),
+				zap.String("sctx", objectStr(s.sctx)),
+				zap.String("caller", objectStr(caller)),
+				zap.String("owner", objectStr(s.owner)),
+				zap.Stack("stack"),
+			)
+			s.avoidReuse = true
+			panic(r)
+		}
+	}, nil
+}
+
+// OwnerClose indicates to close the session.
+// It receives a caller with the type sessionOwner as the argument,
+// and if the caller is not the owner, or it is already closed, it will return without doing anything.
+func (s *session) OwnerClose(caller sessionOwner) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if caller != nil && caller == s.owner {
+		// only close the session when the caller is the owner
+		s.doClose(s.seqInc())
+	}
+}
+
+// Close closes the session without checking the owner.
+func (s *session) Close() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.doClose(s.seqInc())
+}
+
+// doClose closes the session.
+// It should be called with the protection of the mutex.
+func (s *session) doClose(seq uint64) {
+	if s.owner == nil {
+		return
+	}
+
+	defer func() {
+		s.owner = nil
+		s.sctx.Close()
+	}()
+
+	if s.owner != nil {
+		if err := s.owner.onResignOwner(s.sctx); err != nil {
+			s.reportError(
+				"error occurs when resigning the owner",
+				zap.Uint64("seq", seq),
+				zap.Error(err),
+			)
+		}
+	}
+
+	if s.inuse > 0 {
+		s.reportError(
+			"session owner is still in use when closing",
+			zap.Uint64("inuse", s.inuse),
+			zap.Uint64("seq", seq),
+		)
+	}
+}
+
+// checkNotInuse returns an error if the session is still in use.
+// It should be called with the protection of the mutex.
+func (s *session) checkNotInuse(errMsgPrefix string) error {
```

Comment:
```suggestion
func (s *session) checkNotInuseWithoutLock(errMsgPrefix string) error {
```

I think this would make it clearer to the caller that they need to acquire the lock before calling this function.